modules/ssl/ssl_engine_pphrase.c

```diff
@@ -873,6 +877,133 @@ apr_status_t modssl_load_engine_keypair(server_rec *s, apr_pool_t *p,
     ENGINE_free(e);
 
     return APR_SUCCESS;
+}
+#endif
+
+#if MODSSL_USE_OPENSSL_STORE
+static apr_status_t modssl_load_keypair_store(server_rec *s, apr_pool_t *p,
+                                              const char *vhostid,
+                                              const char *certid,
+                                              const char *keyid,
+                                              X509 **pubkey,
+                                              EVP_PKEY **privkey)
+{
+    OSSL_STORE_CTX *sctx;
+    UI_METHOD *ui_method = get_passphrase_ui(p);
+    pphrase_cb_arg_t ppcb;
+
+    memset(&ppcb, 0, sizeof ppcb);
+    ppcb.s = s;
+    ppcb.p = p;
+    ppcb.bPassPhraseDialogOnce = TRUE;
+    ppcb.key_id = vhostid;
+    ppcb.pkey_file = keyid;
+
+    *privkey = NULL;
+    sctx = OSSL_STORE_open(keyid, ui_method, &ppcb, NULL, NULL);
+    if (!sctx) {
+        ap_log_error(APLOG_MARK, APLOG_EMERG, 0, s, APLOGNO(10131)
+                     "Init: OSSL_STORE_open failed for private key identifier `%s'",
+                     keyid);
+        return ssl_die(s);
+    }
+
+    while (!OSSL_STORE_eof(sctx)) {
+        OSSL_STORE_INFO *info = OSSL_STORE_load(sctx);
+        if (!info) {
+            ap_log_error(APLOG_MARK, APLOG_EMERG, 0, s, APLOGNO(10131)
+                         "Init: OSSL_STORE_load failed for private key identifier `%s'",
+                         keyid);
+            OSSL_STORE_close(sctx);
+            return ssl_die(s);
+        }
+
+        switch (OSSL_STORE_INFO_get_type(info)) {
+        case OSSL_STORE_INFO_PKEY:
+            *privkey = OSSL_STORE_INFO_get1_PKEY(info);
+            break;
+        default:
+            OSSL_STORE_INFO_free(info);
+            continue;
+        }
+
+        OSSL_STORE_INFO_free(info);
+        break;
+    }
+
+    OSSL_STORE_close(sctx);
+
+    if (!*privkey) {
+        ap_log_error(APLOG_MARK, APLOG_EMERG, 0, s, APLOGNO(10131)
+                     "Init: OSSL_STORE_INFO_PKEY lookup failed for private key identifier `%s'",
+                     keyid);
+        OSSL_STORE_close(sctx);
+        return ssl_die(s);
+    }
+
+    if (certid) {
+        *pubkey = NULL;
```

Comment:
This could go with the early `*privkey = NULL;` above for the error cases and/or !certid. Maybe there should a cleanup code for `*privkey` and `*pubkey` if not NULL and an error happens at some point (`goto cleanup` in a common pattern in httpd), unless it's a fatal error and httpd is about to exit anyway.

Reply:
OK, will move op the `*pubkey = NULL;`
All error cases use `ssl_die()` so I guess a cleanup is not needed. 
It was that way for the engine case already, so I guess it really should die when it can't load the key/cert for any reason.